### ISSUE TITLE
Alternate de-uglified URLs, as "/file" (no extension), with "/dir/" (with trailing slash) only for directory-like pages

### ISSUE THREAD

**Issue by chrish42** [gohugoio/hugo#7458]
For me, what looks cleanest for a website's URL structure is when the leaf pages are mapped to `/some-file` (no trailing slash) but without a `.html` extension, and `/foo/` is used only for pages that are directory-like. For me, this is the cleanest design for URLs. This works really well together with hosting on a S3, etc. bucket, where you can push metadata to set the correct content-type when there is no extension for the file.

I'm willing to do the work and create a pull request to support this, but before I start coding, it'd like to gauge the openness to such a change, so as to avoid doing all the work to then have it rejected as not a desirable feature.

For Hugo, this would mean:

1. An alternate way of "de-uglifying" URLs that uses `/foo/` only for "directory-like" pages, but that...
2. Keeps note of the content-type of files where the extension is stripped (so we can pass those to the program that uploads to the S3, etc. bucket) and create a content-type map file with this info, optionally
3. And a config switch to access this new functionality.

An alternate way for point 2 might be to restrict `/foo` only for HTML pages (likely what will happen in practice anyways) and then any extension-less file can be assumed to be one. Also, for some tools, the content-type map may not be even needed. At least for GCP, `gsutil` has a mode that auto-detects content-type from the content of the file. I'm guessing it has no problem at all figuring out which files are HTML (from the `<!HTML ...>` header.

Thoughts? Is this something that Hugo would consider supporting, if a pull request implementing this were to magically appear? :-)

**Comment by unor** [comment]
### See also

Issues/PRs:

* https://github.com/gohugoio/hugo/issues/492
* https://github.com/gohugoio/hugo/pull/3556
* https://github.com/gohugoio/hugo/pull/3934
* https://github.com/gohugoio/hugo/issues/5098

Threads:

* https://discourse.gohugo.io/t/links-without-a-trailing-slash-a-solution/6760
* https://discourse.gohugo.io/t/links-without-a-trailing-slash/831
* https://discourse.gohugo.io/t/hugo-support-for-urls-without-a-trailing-slash/6763

Blog posts:

* https://www.ntaso.com/hugo-trailing-slashes/
* https://kevinsweet.com/pretty-hugo-urls-without-a-trailing-slash

**Comment by chrish42** [comment]
Still relevant for me, given the multiple times this has been requested, etc. 

For this to work, it needs to be combined with something with something that sets the right mime-type on the extension-less files uploaded for the webpages, so it needs some support from the webserver and shouldn't be the default behavior. But, at least for folks using static hosting on AWS, GCP, etc., support for per-file mime-types is built-in. I feel like it would be a nice addition.

Basically:
* Upload directories as directories, files as files, according to the URL path.
* At least for the files intended to be accessed by users directly via a URL, they are uploaded without an extension, but with their mime-type set correctly.
